keps/1714-fair-sharing/README.md

```diff
@@ -0,0 +1,375 @@
+
+* [S1] A sub-organization to which CQ x belongs (its cohort, parent cohort, or CQ itself)
+ still has enough resources to admit the workload within its nominal quota, as the total
+ amount of its workload consumption is lower than the quota. However, as there are not
+ enough resources in the system, someone must be borrowing resources. Thus, if we find
+ a potential borrower, we can consider preempting some of its workloads, up to the point
+ when it stops borrowing. This is equivalent to “reclaimWithinCohort” in ClusterQueues.
+
+
+* [S2] A sub-organization to which CQ x belongs is borrowing, however it seems that it is
+borrowing too little compared to other sub-organizations that are also borrowing, so some
+action is needed to enforce fair sharing. We should compare how much the sub-orgs are
+borrowing compared to each other, and preempt some workloads up to the point when its fair
+share would be smaller than the sub-org for which preemptions are executed.
+
+Let’s formalize the above description.
+
+By LCA(x,y) we define the lowest common ancestor of two ClusterQueues x and y in the cohort
+hierarchy. AlmostLCA(x,y) is the last but one element on the path from x to LCA(x,y) in the cohort hierarchy.
+
+When trying to preempt some workloads to admit workload w to CQ x we consider all other
+CQs that are in the same hierarchy. For each of such CQ y we check whether scenario S1
+or S2 occur. To do that we compare sub-organizations that are rooted in AlmostLCA(x,y) and AlmostLCA(y,x).
+
+![LCA diagram](lca.png)
+
+First of all we validate if y is not using its local nominal quota, somewhere on the path.
+If y was using some local nominal quota that lies entirely under AlmostLCA(y,x) then x should
+not be allowed to preempt any of its workload. So we check whether both y and all cohorts on
+the way up to AlmostLCA(y,x) are borrowing. If that happens then y may have something to
+do with x being unable to admit the workload. 
+
+We trigger scenario S1 if AlmostLCA(x,y) is not borrowing but x still has too little
+resources. We put all y’s workloads on a list of potential preemptions, out of which
+we will later (once we check all other CQs) select one that is most reasonable. 
+
+
+We trigger scenario S2 if AlmostLCA(x,y) is borrowing. We check whether AlmostLCA(x,y)
+and AlmostLCA(y,x) are borrowing roughly the same amount. To do this we use a value 
+function (to be defined later) that returns a value of the borrowed capacity.
+
+#### Choosing workloads from ClusterQueues for preemption
+
+For each workload z in y we check whether if:
+
+[S2-a] value of AlmostLCA(y,x) without z is still higher than value of AlmostLCA(x,y)
+with admitted workload x. Y’s sub-orgs will still be better than X’s sub-org after we
+preempt z and admit w, thus z is a reasonable candidate to re-balance fair sharing.
+
+
+[S2-b] value of AlmostLCA(y,x) is lower than AlmostLCA(x,y) with admitted workload x.
```

Comment:
this seems to be a typo
```suggestion
[S2-b] value of AlmostLCA(y,x) is strictly higher than AlmostLCA(x,y) with admitted workload x.
```

Also clarify that this implies that, using this rule, the maximum share value in the cluster monotonically decreases, eventually stopping.